Core: Memory Fixes

StevenMiller123 opened the pull request:
This PR makes our memory code more robust, preventing a variety of exceptions and strange behavior I observed when using my recently shared memory stress test homebrew on shadPS4.

- Fixes `OrbisVirtualQueryInfo` struct. My homebrew was reading garbage data with the BitFields and array in place. I've replaced the BitField class use with C-style bitfields, and the std::array with a primitive char array instead.
- Proper low address handling in `sceKernelVirtualQuery`. Our previous code didn't check if the inputted address was below our vma map, which led to OOB memory accesses. My edge case error returns are based on hardware tests here.
- Edge case handling for `SearchFree`. Our previous code encountered unpredictable behavior if the inputted address was too high, or if there were no free memory areas to map.
- Changed default nameless memory mapping behavior. On real hardware, nameless mappings will have the name "anon" followed by the code address responsible for the memory call. For simplicity sake, I just used "anon" as the default name instead.
- Fixes a variety of broken memory asserts, caused by a fundamental misunderstanding of how `FindVMA` and `FindDmemArea` behave.
- Fixes a long unnoticed regression caused by #2080 
- Fixes an issue where `MapMemory` would assert when the area to map was only partially reserved.

With these fixes, my memory homebrew now executes all tests on shadPS4. Many of the returned results are still inaccurate, but these are due to deeper issues that I believe are out of scope for this PR.

Currently:
- Regresses EA SPORTS™ UFC® (CUSA00264) 
     - The only reason this game works in main is due to a bug in our code. Fixing this bug broke the game, and attempting to properly fix the regression broke Windows support.
- Improves Evolve (CUSA00432)
- Improves Cars 3: Driven to Win (CUSA07027)
- Improves Shadow of the Tomb Raider (CUSA10872)
- Improves The Order: 1886 (CUSA00076)
- Improves Final Fantasy XV (CUSA01633)

Commit:
Fix VirtualQuery behavior on low addresses.

Commit:
Fix VirtualQuery struct

Somewhere in our BitField and array use, the size of our VirtualQuery struct became larger than the struct used on real hardware.
Fixing this fixes some data corruption visible in the name parameter during my tests.

Commit:
Default name to anon

On real hardware, nameless mappings are given the name "anon:address" where address appears to be the address that made the memory call.
For simplicity sake, I'll stick to the name "anon" for now.

Commit:
Place an upper bound on returns from SearchFree

Right now, this upper bound is set based on the limitations of our GPU buffer cache and page table.
Someone with more experience in that area of code should probably fix that at some point.

Commit:
More anons

Commit:
Clang

squidbus commented on src/core/memory.cpp:
```diff
@@ -585,12 +609,14 @@ int MemoryManager::VirtualQuery(VAddr addr, int flags,
+    info->is_pooled = vma.type == VMAType::PoolReserved || vma.type == VMAType::Pooled ? 1 : 0;
+    info->is_committed = vma.IsMapped() ? 1 : 0;
+
+    strcpy(info->name, vma.name.data());
```
Prefer `strncpy` with maximum length of `info->name`.

StevenMiller123 replied:
Should be resolved now.

StevenMiller123 committed:
Fix name in sceKernelMapNamedDirectMemory

StevenMiller123 committed:
strncpy instead of strcpy

Hardcoded the constant size for now, I need to review how real hardware behaves here to determine if anything else is necessary for this to be accurate.

Commit:
Fix name behavior

All memory naming functions restrict the name size to a 31 character limit, and return `ORBIS_KERNEL_ERROR_ENAMETOOLONG` if that limit is exceeded.
Since this value is constant for all functions involving names, I've defined it as a constant in kernel's memory.h, and used that in place of any hardcoded 32 character limits.

Commit:
Error logging

Hopefully this helps in catching the UFC regression?

Commit:
Increase address space upper bound

Probably needs heavy testing, especially on Mac/Windows.
This increases the address space, as needed to accommodate strange memory behaviors seen in UFC.

Commit:
VirtualQuery fix

Due to limitations of certain platforms, we initialize our vma_map with 3 separate free mappings.
As such, we need to use a while loop here to accurately query mappings with high addresses

Commit:
Fix mappings to high addresses

The PS4's GPU can only handle 40bit addresses. Our texture cache and buffer cache were designed around these limits, and mapping to higher addresses would cause segmentation faults and access violations.
To fix these crashes, only map to the GPU if the mapping is fully contained within the address space the GPU should access.
I'm open to suggestions on how to make this cleaner

Commit:
Revert "Increase address space upper bound"

This reverts commit 3d50eeeebb6aa40e38d6f87e6480235c917843f3.

StevenMiller123 committed:
Revert VirtualQuery while loop

Windows wasn't happy with this, again.
Will try to debug and properly fix this when I have a good chance.

StevenMiller123 committed:
Merge branch 'shadps4-emu:main' into memory-fixes-again

Commit:
Fix asserts

FindVMA, due to the way it's programmed, never actually returns vma_map.end(), the furthest it ever returns is the last valid memory area. All those asserts we involving vma_map.end() never actually trigger due to this.
This commit removes redundant asserts, adds messages to asserts that were lacking them, and fixes all asserts designed to detect out of bounds memory accesses so they actually trigger.
I've also fixed some potential memory safety issues.

Commit:
Proper error behavior in QueryProtection

Might as well handle this properly while I'm here.

Commit:
Clang

StevenMiller123 committed:
More information about ReserveVirtualRange results

Should help debug issues like the one in The Order: 1886 (CUSA00076)

Commit:
Fix assert message

Commit:
Update assert message

Extra space

Commit:
Fix my bug

Oh hey, finally something that's my fault.

Commit:
Fix rasterizer unmaps

Should use adjusted_size here, otherwise we could unmap too much.
Thanks to diegolix29 for spotting this.

Commit:
Fix edge case in MapMemory

Code comments explain everything.
This should fix some memory asserts.

Commit:
Fix fix

Avoid running the code path if it's unnecessary, since there are many additional edge cases to handle when the VMA map is small.

StevenMiller123 committed:
Fix fix fix

Should prevent infinite loops, haven't tested properly yet though.

squidbus commented on src/core/libraries/kernel/memory.cpp:
```diff
@@ -126,17 +126,11 @@ s32 PS4_SYSV_ABI sceKernelAvailableDirectMemorySize(u64 searchStart, u64 searchE
 
 s32 PS4_SYSV_ABI sceKernelReserveVirtualRange(void** addr, u64 len, int flags, u64 alignment) {
-    LOG_INFO(Kernel_Vmm, "addr = {}, len = {:#x}, flags = {:#x}, alignment = {:#x}",
-             fmt::ptr(*addr), len, flags, alignment);
```
If this is moved to capture the out address, would prefer to keep this here and then log the out address separately, so we can see this in any error case that we may not reach the end.

StevenMiller123 replied:
Fixed with my newest commit.

Commit:
Split logging for inputs and out_addr in ReserveVirtualRange

Addresses review comments.
